pydantic/_internal/_model_construction.py

```diff
@@ -202,6 +205,11 @@ def build_inner_schema(  # noqa: C901
                 raise
         raise PydanticUndefinedAnnotation(name) from e
 
+    # Also raise PydanticUndefinedAnnotation if a type_hint is a ForwardRef type (i.e. not yet defined)
+    for t in type_hints.values():
+        if isinstance(t, typing.ForwardRef):
+            raise PydanticUndefinedAnnotation(t.__forward_arg__)
```

Comment:
I don't see a test which correctly raises this error.

In other words, if I were to check out your branch and remove this code (or change it to raise no error), I would want a test to fail.

Reply:
@ybressler when I comment out this block on my dev system (Python 3.10), the two tests fail:

- test_undefined_type_raises_warning_if_forward_ref
- test_disable_undefined_types_warning_if_missing_class_is_a_forwardref

This was the reason I added that block of code.  This makes me wonder if there are different behaviors on different Python versions or systems, since the importance of this block of code was initially questioned by you and @samuelcolvin.

Ah, yes... I think I found the source of the problem/difference:

Inside `_typing_extra.get_type_hints()` it checks if Python >= 3.10 and then uses:

Python >= 3.10 -> the built-in `typing.get_type_hints()` for Python >= 3.10
Python < 3.10 -> a special implementation with partially copied code from 3.10

_typing_extra.py lines 216-234
```python
if sys.version_info >= (3, 10):  # noqa C901
    get_type_hints = typing.get_type_hints

else:
    """
    For older versions of python, we have a custom implementation of `get_type_hints` which is a close as possible to
    the implementation in CPython 3.10.8.
    """
    fr_has_is_class = True

    def ForwardRefWrapper(arg: Any, is_argument: bool = True, *, is_class: bool = False) -> typing.ForwardRef:
        """
        Wrapper for ForwardRef that accounts for the `is_class` argument missing in older versions.
        The `module` argument is omitted as it breaks <3.9 and isn't used in the calls below.

        See https://github.com/python/cpython/pull/28560 for some background

        Implemented as EAFP with memory.
        """
```

Untangling this is probably important and is something I will defer to others :)

It's great that you noticed this @ybressler. My guess is that this will make sense to @samuelcolvin since the custom implementation of `get_type_hints` appears to be his work.

Reply:
Agreed. Let's wait till Samuel reviews. I imagine a suitable approach is to create a follow up issue (perhaps backlogged) to untangle this functionality. Meanwhile, code is good as is.